File: leetcode/typescript/1427.ts
function stringShift(s: string, shift: number[][]): string {
  let netShift = 0;
  for (const [a, b] of shift) {
    netShift += a === 0 ? -b : b;
  }
  netShift %= s.length;
  if (netShift < 0) {
    netShift += s.length;
  }
  return s.slice(-netShift) + s.slice(0, -netShift);
}

console.log(
  stringShift("abc", [
    [0, 1],
    [1, 2],
  ])
);

(() => {
  const testCases = [
    {
      name: "Regular dominoes",
      func: stringShift,
      tests: [
        {
          s: "abc",
          shift: [
            [0, 1],
            [1, 2],
          ],
          expected: "cab",
        },
        {
          s: "abcdefg",
          shift: [
            [1, 1],
            [1, 1],
            [0, 2],
            [1, 3],
          ],
          expected: "efgabcd",
        },
      ],
    },
  ];

  testCases.forEach((testCase) => {
    console.log(`Testing ${testCase.name}:`);
    testCase.tests.forEach((test, index) => {
      const result = testCase.func(test.s, test.shift);
      const passed = result === test.expected;
      console.log(`Test ${index + 1}: ${passed ? "PASSED" : "FAILED"}`);
      console.log(`  Input: tops = [${test.s}], bottoms = [${test.shift}]`);
      console.log(`  Expected: ${test.expected}, Got: ${result}`);
      if (!passed) {
        console.log(`  ❌ Test failed!`);
      }
      console.log();
    });
    console.log();
  });
})();

File: leetcode/go/1427.go
//go:build 1427
// +build 1427

package main

import "fmt"

func stringShift(s string, shift [][]int) string {
	netShift := 0
	for _, op := range shift {
		if op[0] == 0 {
			netShift -= op[1]
		} else {
			netShift += op[1]
		}
	}

	netShift %= len(s)

	if netShift < 0 {
		netShift += len(s)
	}

	return s[len(s)-netShift:] + s[:len(s)-netShift]
}

func main() {
	testCases := []struct {
		s     string
		shift [][]int
		want  string
	}{
		{"abc", [][]int{{0, 1}, {1, 2}}, "cab"},
		{"abcdefg", [][]int{{1, 1}, {1, 1}, {0, 2}, {1, 3}}, "efgabcd"},
		{"abcdefg", [][]int{{1, 1}, {1, 1}, {0, 2}, {1, 3}, {0, 7}}, "efgabcd"},
	}

	for i, tc := range testCases {
		got := stringShift(tc.s, tc.shift)
		if got != tc.want {
			fmt.Printf("Test case %d failed. Got: %s, Want: %s\n", i+1, got, tc.want)
		} else {
			fmt.Printf("Test case %d passed.\n", i+1)
		}
	}
}
